File: javascript/union_find/weighted_quick_union_uf.js
/**
 * A Quick-Union implementation with the following enhancements:
 * 1. Weighted Quick-Union: to avoid tall trees
 * 2. Quick-Union with Path Compression: to reduce node-travel
 */


class WeightedQuickUnionUF {

	/**
	 * Generates a UnionFind data-structure os length `size`
	 * @param {*} size The number of nodes in the UnionFind structure.
	 */
	constructor(size) {
		this.id_ = new Array(size);
		this.size_ = new Array(size);

		// Set id of each node to itself & the size to 1.
		for (let i=0; i < size; i++) {
			this.id_[i] = i;
			this.size_[i] = 1;
		}
	}

	/**
	 * Check if it is a valid node
	 * @param {Number} i The node to check
	 * @returns {Boolean}
	 */
	isValid(i) {
		return (i >= 0 && i < this.id_.length) ? true : false;
	}

	/**
	 * Get the root node of a given node.
	 * @param {Number} i The given node
	 * @returns {Number} The root node
	 */
	root(i) {
		if (!this.isValid(i)) {
			return -1;
		}

		while (i !== this.id_[i]) {
			this.id_[i] = this.id_[this.id_[i]];	// One-pass path compression
			i = this.id_[i];						// Goto root node
		}
		return i;
	}

	/**
	 * Check whether the nodes `p` & `q` are connected.
	 * @param {Number} p First node
	 * @param {Number} q Second node
	 * @returns {Boolean} True if p & q are connected; false otherwise.
	 */
	connected(p, q) {
		return this.root(p) === this.root(q);
	}

	/**
	 * Perform a union of nodes p & q.
	 * @param {Number} p The first node.
	 * @param {Number} q The second node.
	 */
	union(p, q) {
		if (!(this.isValid(p) && this.isValid(q))) {
			return false;
		}

		let proot = this.id_[p];
		let qroot = this.id_[q];

		// Change the root of the smaller tree to point to the root of teh larger tree
		if (this.size_[proot] < this.size_[qroot]) {
			this.id_[proot] = qroot;
			this.size_[qroot] += 1;
		} else {
			this.id_[qroot] = proot;
			this.size_[proot] += 1;
		}
	}

	/**
	 * Print the unions
	 */
	print() {
		var tmp = {};

		this.id_.forEach((p, i) => {
			let r = this.root(i);
			if (r in tmp) {
				tmp[r].push(i);
			} else {
				tmp[r] = [i]
			}
		});

		let out = "";
		for (let i in tmp) {
			out += (out && ' ') + "(" + tmp[i] + ")";
		}

		console.log(out);
	}
}

module.exports = { WeightedQuickUnionUF };

File: java/union_find/WeightedQuickUnionUF.java
package union_find;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;


/**
 * A Quick-Union implementation with the following enhancements:
 * 1. Weighted Quick-Union: to avoid tall trees
 * 2. Quick-Union with Path Compression: to reduce node-travel
 */

public class WeightedQuickUnionUF {
    private int[] id;
    private int[] size;     // IMPROVEMENT #1: Caching size(weight) of each tree
	private int N;

    /**
     * Constructor to create a UnionFind data structure.
     * @param _N Size of the UnionFind data structure.
     */
    public WeightedQuickUnionUF(int _N)
    {
		N = _N;
        id = new int[N];
        size = new int[N];
        // Set id of each object to itself & their size to 1
        // (2*N array accesses)
        for (int i = 0; i < N; i++) {
            id[i] = i;
            size[i] = 1;
        }
    }

    /**
     * Find the root of the given node
     * @param i The given node
     * @return The root node of the given node
     */
    private int root(int i)
    {
        // chase parent pointers until root is reached (depth of array accesses)
        while (i != id[i]) {
            // IMPROVEMENT #2: Simple one-pass flattening of the tree.
            // This keeps the tree completely flat for faster root finding.
            id[i] = id[id[i]];
            i = id[i];
        }
        return i;
    }

    /**
     * Check whether the nodes `p` & `q` are connected
     * @param p First node
     * @param q Second node
     * @return true if p & q are connected; false otherwise.
     */
    public boolean connected (int p, int q)
    {
        // check whether p and q have same root (depth of p and q array accesses)
        return root(p) == root(q);
    }

    /**
     * Perform a union of nodes p & q by connecting these nodes
     * @param p First node
     * @param q Second node
     */
    public void union (int p, int q)
    {
        int proot = root(p);
        int qroot = root(q);
        // IMPROVEMENT #1:
        // change root of smaller tree to point to root of the larger tree (depth of p and q array accesses)
        if (size[proot] < size[qroot]) {
            id[proot] = qroot;
            size[qroot] += size[proot];
        } else {
            id[qroot] = proot;
            size[proot] += size[qroot];
        }
    }

	/**
	 * Print the UnionFind sets
	 */
	public void print()
	{
		Map<Integer, List<Integer>> map = new HashMap<Integer, List<Integer>>();
		for (int i = 0; i < N; i++) {
			int root = root(i);
			map.putIfAbsent(root, new ArrayList<Integer>());
			map.get(root).add(i);
		}

		System.out.println("Union-find sets: ");
		String _out = "";
		for (Entry<Integer, List<Integer>> ee : map.entrySet()) {
			int key = ee.getKey();
			List<Integer> values = ee.getValue();
			_out += (_out.length() > 0 ? " " : "") + "(" + key;
			for (int i = 0; i < values.size(); i++)
				if (values.get(i) != key)
					_out += (", " + values.get(i));
			_out += ")";
		}

		System.out.println(_out);
	}
}
